src/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs

```diff
@@ -879,6 +1042,34 @@ public static MethodCallExpression Call(Type type, string methodName, Type[] typ
         ///<paramref name="instance" />.Type is not assignable to the declaring type of the method represented by <paramref name="method" />.-or-The number of elements in <paramref name="arguments" /> does not equal the number of parameters for the method represented by <paramref name="method" />.-or-One or more of the elements of <paramref name="arguments" /> is not assignable to the corresponding parameter for the method represented by <paramref name="method" />.</exception>
         public static MethodCallExpression Call(Expression instance, MethodInfo method, IEnumerable<Expression> arguments)
         {
+            IReadOnlyList<Expression> argumentList = arguments as IReadOnlyList<Expression>;
+
+            if (argumentList != null)
+            {
+                switch (argumentList.Count)
+                {
+                    case 0:
+                        return Call(instance, method);
+                    case 1:
+                        return Call(instance, method, argumentList[0]);
+                    case 2:
+                        return Call(instance, method, argumentList[0], argumentList[1]);
+                    case 3:
+                        return Call(instance, method, argumentList[0], argumentList[1], argumentList[2]);
+                }
+
+                if (instance == null)
```

Comment:
Correct, stepping on the public Arguments property causes the magic morphing into a collection. Some external visitors (a la a ExpressionVisitor<T> we built here) step on that frequently, but the recent change where IArgumentProvider was made public (not sure exactly when that happened) makes it possible to avoid the materialization and pick out the arguments without causing the mutation to a collection to happen.